src_kernels/cuda/grid_update.rs

```diff
@@ -125,25 +126,29 @@ fn update_single_cell(
 
                         if apply_friction {
                             let normal_vel = cell_velocity.dot(&normal);
+                            let dist_with_margin = (dist - cell_width * 0.5).max(0.0);
 
-                            if normal_vel < 0.0 {
-                                let dist_with_margin = dist - cell_width;
-                                if is_inside || dist_with_margin <= 0.0 {
-                                    let tangent_vel =
-                                        cell_velocity - normal_vel * normal.into_inner();
-                                    let tangent_vel_norm = tangent_vel.norm();
-
-                                    cell_velocity = tangent_vel;
-
-                                    if tangent_vel_norm > 1.0e-10 {
-                                        let friction = collider.friction;
-                                        cell_velocity = tangent_vel / tangent_vel_norm
-                                            * (tangent_vel_norm + normal_vel * friction).max(0.0);
-                                    }
-                                } else if -normal_vel * dt > dist_with_margin {
-                                    cell_velocity -=
-                                        (dist_with_margin / dt + normal_vel) * normal.into_inner();
-                                }
+                            if -normal_vel * dt > dist_with_margin {
+                                // NOTE: if we enter this code, normal_vel is negative.
+                                let tangent_vel = cell_velocity - normal_vel * normal.into_inner();
+                                let tangent_vel_norm = tangent_vel.norm();
+
+                                let new_normal_vel = normal_vel.max(-dist_with_margin / dt);
```

Comment:
Even if we zero the normal velocity out completely, particles can still approach the boundary as their non-zero-kernel space goes further. We may want to extend the margin s.t. we cover the whole range of an approaching particle.

Reply:
This amounts to changing the line above to `let dist_with_margin = (dist - cell_width * 1.5).max(0.0);`.

This also restores the previous behavior in the fluid example btw.